src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/tools/monitoring/outputMonitor.ts

```diff
@@ -0,0 +1,254 @@
+export class OutputMonitor extends Disposable implements IOutputMonitor {
+	private _isIdle = false;
+	private _state: OutputMonitorState = OutputMonitorState.Initial;
+	get state(): OutputMonitorState {
+		return this._state;
+	}
+
+	private readonly _onDidFinishCommand = this._register(new Emitter<void>());
+	readonly onDidFinishCommand = this._onDidFinishCommand.event;
+	private readonly _onDidIdle = this._register(new Emitter<void>());
+	readonly onDidIdle = this._onDidIdle.event;
+	private readonly _onDidTimeout = this._register(new Emitter<void>());
+	readonly onDidTimeout = this._onDidTimeout.event;
+
+	get isIdle(): boolean {
+		return this._isIdle;
+	}
+
+	constructor(
+		private readonly _execution: IExecution,
+		@ILanguageModelsService private readonly _languageModelsService: ILanguageModelsService,
+		@ITaskService private readonly _taskService: ITaskService,
+		@IChatWidgetService private readonly _chatWidgetService: IChatWidgetService,
+		private readonly _pollFn?: (execution: IExecution, token: CancellationToken, taskService: ITaskService) => Promise<IPollingResult | undefined> | undefined,
+	) {
+		super();
+	}
+
+	async startMonitoring(
+		chatService: IChatService,
+		command: string,
+		invocationContext: any,
+		token: CancellationToken
+	): Promise<IPollingResult & { pollDurationMs: number }> {
+
+		const pollStartTime = Date.now();
+
+		let result = await this._pollForOutputAndIdle(this._execution, false, token, this._languageModelsService, this._chatWidgetService, this._taskService, this._pollFn);
+
+		if (this._state === OutputMonitorState.Timeout) {
+			result = await this._racePollingOrPrompt(
+				() => this._pollForOutputAndIdle(this._execution, true, token, this._languageModelsService, this._chatWidgetService, this._taskService, this._pollFn),
+				() => this._promptForMorePolling(command, token, invocationContext, chatService),
+				result,
+			);
+		}
+
+		return { ...result, pollDurationMs: Date.now() - pollStartTime };
+	}
+
+	/**
+	 * Waits for either polling to complete (terminal idle or timeout) or for the user to respond to a prompt.
+	 * If polling completes first, the prompt is removed. If the prompt completes first and is accepted, polling continues.
+	 */
+	private async _racePollingOrPrompt(
+		pollFn: () => Promise<IRacePollingOrPromptResult>,
+		promptFn: () => Promise<{ promise: Promise<boolean>; part?: Pick<ChatElicitationRequestPart, 'hide' | 'dispose'> }>,
+		originalResult: IPollingResult,
+	): Promise<IRacePollingOrPromptResult> {
+		type Winner =
+			| { kind: 'poll'; result: IRacePollingOrPromptResult }
+			| { kind: 'prompt'; continuePolling: boolean };
+
+		const { promise: promptP, part } = await promptFn();
+
+		const pollPromise = pollFn().then<Winner>(result => ({ kind: 'poll', result }));
+		const promptPromise = promptP.then<Winner>(continuePolling => ({ kind: 'prompt', continuePolling }));
+
+		let winner: Winner;
+		try {
+			winner = await Promise.race([pollPromise, promptPromise]);
```

Comment:
Does the other one need to be canceled when there's a winner?

Reply:
No because if the prompt is the winner, that means the user accepted or rejected the prompt and it's handled there